docs/RationaleSeq.md

```diff
@@ -0,0 +1,108 @@
+
+This document describes various design points of the `seq` dialect, why it is
+the way it is, and current status. This follows in the spirit of other [MLIR
+Rationale docs](https://mlir.llvm.org/docs/Rationale/).
+
+## Introduction
+
+Digital logic is generally split into two categories: combinational and
+sequential. CIRCT contains a `comb` dialect to model the basic combinational
+operations and the (future) `seq` dialect which is discussed here. The
+intention of the `seq` dialect is to provide a set of stateful constructs
+which can be used to model sequential logic, independent of the output method
+(e.g. SystemVerilog).
+
+We have yet to flesh this dialect out to represent all common stateful
+components. We start with a register and will build from there.
+
+## Definitions
+
+For the sake of precision, we use the following definitions:
+
+- Physical devices:
+  - **Unclocked Latch:** A memory element which is only sensitive to the
+  levels of its inputs. Has no clock. Example: SR Latch.
+  - Clocked (gated) **latch:** A latch wherein the inputs are gated by a
+  clock. Transparent the entire time the clock is high. Generally referred to
+  as a "latch". Examples: "gated SR latch", "D latch".
+  - Edge-triggered **flip-flop:** An edge-sensitive memory element. Captures
+  the input value on one or both clock edges. Variants: posedge FF, negedge
+  FF, "edge-sensitive" FF (captures the input value on both edges),
+  resettable FF.
+- Abstract models:
+  - **Register:** A synchronous, resettable memory element. Can be
+  implemented using any of the above "circuit level" elements.
+
+## The computational register operation
+
+The `seq.compreg` op models an abstract notion of a "register", independent
+of its implementation (e.g. latch, D flip-flop). This specific register op is
+intended to support "computation support" or "reset-agnostic code" and thus
+it cannot be used to model all the behaviors of a SystemVerilog register.
+(E.g. FSM and pipeline registers.) It is intended to be 'lowered' to a
+specific implementation, which may be an `sv.always_ff` block, a device
+primitive instantiation, or even a feedback loop in all `comb` logic.
+
+Our intention is to allow analysis and optimization of sequential logic
+without having to reason about implementation-specific behavior. This makes
+it somewhat distinct from the SystemVerilog register wherein one has to
+account for details like reset behavior and implementation (latch vs.
+flip-flop).
+
+`Reg` has four operands:
+
+- **input**: The value to be captured at 'clock'. Generally called 'd'.
+Accepts any type, results in the same type. Does not support any notion of
+addressing, meaning that this operation sets / reads the entire value.
+- **clock**: Capture 'value' on the positive edge of this signal.
+- **reset**: Signal to set the state to 'resetValue'. Optional.
+- **resetValue**: A value which the state is set to upon reset. Required iff
+'reset' is present.
+
```

Comment:
If we are considering treading the "async vs sync" behavior of the reset as a higher level concept, why even capture the reset wire, either? Or the clock? I could certainly see value in separating out the "stateful logic to do the thing you want to do" from the "clock and reset domain" considerations. As it is right now it feels like we are straddling the line in an awkward way.

Reply:
I've previously proposed the "clock/reset" regions with implicit clock and reset signals. That's too high level, but I want to add it to a higher level dialect at some point. It makes lowering to this op difficult for all of the existing dialects to lower to -- they all expect a clock operand. The reason to capture the reset wire is that not all registers have to be reset -- as a rule of thumb, only the control regs have to be.

> As it is right now it feels like we are straddling the line in an awkward way.

I'm not saying that this should be unspecified forever, I'm just saying that we shouldn't design that specification method *now*. I'd rather get in a very simple MVP and get some experience with using it / analyzing it before we decide how to model resets if they prove necessary for this particular op.

I suspect that a more general register op will be necessary (modeling different clock types and reset styles) but that this one will be useful on its own. I may be wrong about that.

If we have time tomorrow, I think we should continue discussing this in the ODM.